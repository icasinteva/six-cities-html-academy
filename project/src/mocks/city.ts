import { LocationItem } from '../const';
import {City} from '../types/map';

export const CITY: City = {
  title: LocationItem.Amsterdam,
  lat: 52.3676,
  lng: 4.9041,
};

export const CITIES: City[] = [
  CITY,
  {
    title: LocationItem.Paris,
    lat: 48.8566,
    lng: 2.3522,
  },
];
