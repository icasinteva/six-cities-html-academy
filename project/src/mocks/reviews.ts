import { Reviews } from '../types/offer';

export const reviews: Reviews = [
  {
    author: 'Max',
    rating: 4,
    avatar: 'img/avatar-max.jpg',
    text: 'A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.',
    date: {
      month: 'April',
      year: 2019,
    },
  },
  {
    author: 'Max',
    rating: 4,
    avatar: 'img/avatar-max.jpg',
    text: 'A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.',
    date: {
      month: 'April',
      year: 2019,
    },
  },
  {
    author: 'Max',
    rating: 4,
    avatar: 'img/avatar-max.jpg',
    text: 'A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.',
    date: {
      month: 'April',
      year: 2019,
    },
  },
  {
    author: 'Max',
    rating: 4,
    avatar: 'img/avatar-max.jpg',
    text: 'A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.',
    date: {
      month: 'April',
      year: 2019,
    },
  },
  {
    author: 'Max',
    rating: 4,
    avatar: 'img/avatar-max.jpg',
    text: 'A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.',
    date: {
      month: 'April',
      year: 2019,
    },
  },
];
