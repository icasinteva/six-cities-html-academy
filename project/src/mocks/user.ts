import { AuthorizationStatus } from '../const';

export const generateUser = (authorizationStatus: AuthorizationStatus) => authorizationStatus === AuthorizationStatus.Auth ? {
  id: 1,
  isPro: false,
  name: 'Oliver',
  email: 'Oliver.conner@gmail.com',
  avatarUrl: 'img/avatar-angelina.jpg',
  token: 'qhdowry9283rrw',
} : {
  name: 'Anonim',
};
