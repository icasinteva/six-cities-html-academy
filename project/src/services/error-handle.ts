import request from 'axios';
import { toast } from 'react-toastify';
import { HTTPCode } from '../const';
import { ErrorType } from '../types/error';

export const errorHandle = (error: ErrorType): void => {
  if (!request.isAxiosError(error)) {
    throw error;
  }


  const { response } = error;

  if (response) {
    const { status, data } = response;

    switch (status) {
      case HTTPCode.BadRequest: {
        toast.info(data.error);
        break;
      }
      case HTTPCode.Unauthorized: {
        toast.info(data.error);
        break;
      }
      case HTTPCode.NotFound: {
        toast.info(data.error);
        break;
      }
    }
  }
};
