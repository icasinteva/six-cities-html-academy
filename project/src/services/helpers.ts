
import { FavoritesByCity, Offer } from '../types/offer';

export const loadOffersByCity = (offers: Offer[], activeCity: string) => offers.filter(({ city }) => city.name === activeCity);

export const getFavorites = (offers: Offer[] = []) => {
  const favorites = offers.filter(({ isFavorite }) => isFavorite);
  const favoritesByCity: FavoritesByCity = {};

  for (const data of favorites) {
    const city = data.city.name;
    const cityFavorites = favoritesByCity[city];

    if (!cityFavorites) {
      favoritesByCity[city] = favorites.filter((offer) => offer.city.name === city);
    } else {
      continue;
    }
  }

  return favoritesByCity;
};

export const formatDate = (value: string) => {
  const date = new Date(value);
  const month = date.toLocaleString('default', { month: 'long' });
  const year = date.toLocaleString('default', { year: 'numeric' });

  return `${  month  } ${  year}`;
};
