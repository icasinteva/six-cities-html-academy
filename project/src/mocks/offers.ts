import { LocationItem } from '../const';
import { LocationOffers, Offers } from '../types/offer';
import { reviews } from './reviews';

// const generateId = () => `${Math.floor(Math.random() * 100)}`;
const generateRating = () => Math.floor(Math.random() * 5) + 1;
const generatePrice = () => Math.floor(Math.random() * 355) + 90;

export const offers: Offers = [
  {
    gallery: ['img/apartment-01.jpg', 'img/apartment-02.jpg', 'img/apartment-03.jpg', 'img/apartment-01.jpg', 'img/apartment-02.jpg', 'img/apartment-03.jpg'],
    id: 'Offer-1',
    imageSrc: 'img/apartment-01.jpg',
    premium: true,
    favorite: false,
    facilities: ['Wi-Fi', 'Washing machine', 'Towels', 'Heating', 'Coffee machine', 'Baby seat', 'Kitchen', 'Dishwasher', 'Cabel TV', 'Fridge'],
    lat: 52.3909553943508,
    lng: 4.85309666406198,
    price: generatePrice(),
    rating: generateRating(),
    title: 'Beautiful & luxurious apartment at great location',
    type: 'Apartment',
    host: {
      pro: true,
      name: 'Angelina',
      avatar: 'img/avatar-angelina.jpg',
      description: ['A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.', 'An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the location comes to rest in this alley flowery and colorful.'],
    },
    reviews: [reviews[0]],
  },
  {
    gallery: ['img/apartment-01.jpg', 'img/apartment-02.jpg', 'img/apartment-03.jpg', 'img/apartment-01.jpg', 'img/apartment-02.jpg', 'img/apartment-03.jpg'],
    id: 'Offer-2',
    imageSrc: 'img/apartment-02.jpg',
    favorite: true,
    facilities: ['Wi-Fi', 'Washing machine', 'Towels', 'Heating', 'Coffee machine', 'Baby seat', 'Kitchen', 'Dishwasher', 'Cabel TV', 'Fridge'],
    lat: 52.369553943508,
    lng: 4.85309666406198,
    premium: false,
    price: generatePrice(),
    rating: generateRating(),
    title: 'Beautiful & luxurious apartment at great location',
    type: 'Apartment',
    host: {
      pro: true,
      name: 'Angelina',
      avatar: 'img/avatar-angelina.jpg',
      description: ['A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.', 'An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the location comes to rest in this alley flowery and colorful.'],
    },
    reviews: [reviews[2], reviews[0]],
  },
  {
    gallery: ['img/apartment-01.jpg', 'img/apartment-02.jpg', 'img/apartment-03.jpg', 'img/apartment-01.jpg', 'img/apartment-02.jpg', 'img/apartment-03.jpg'],
    id: 'Offer-3',
    imageSrc: 'img/apartment-03.jpg',
    favorite: true,
    facilities: ['Wi-Fi', 'Washing machine', 'Towels', 'Heating', 'Coffee machine', 'Baby seat', 'Kitchen', 'Dishwasher', 'Cabel TV', 'Fridge'],
    lat: 52.3909553943508,
    lng: 4.929309666406198,
    premium: false,
    price: generatePrice(),
    rating: generateRating(),
    title: 'Beautiful & luxurious apartment at great location',
    type: 'Apartment',
    host: {
      pro: true,
      name: 'Angelina',
      avatar: 'img/avatar-angelina.jpg',
      description: ['A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.', 'An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the location comes to rest in this alley flowery and colorful.'],
    },
    reviews,
  },
  {
    gallery: ['img/apartment-01.jpg', 'img/apartment-02.jpg', 'img/apartment-03.jpg', 'img/apartment-01.jpg', 'img/apartment-02.jpg', 'img/apartment-03.jpg'],
    id: 'Offer-4',
    imageSrc: 'img/apartment-02.jpg',
    premium: true,
    favorite: false,
    facilities: ['Wi-Fi', 'Washing machine', 'Towels', 'Heating', 'Coffee machine', 'Baby seat', 'Kitchen', 'Dishwasher', 'Cabel TV', 'Fridge'],
    lat: 52.3809553943508,
    lng: 4.939309666406198,
    price: generatePrice(),
    rating: generateRating(),
    title: 'Beautiful & luxurious apartment at great location',
    type: 'Apartment',
    host: {
      pro: true,
      name: 'Angelina',
      avatar: 'img/avatar-angelina.jpg',
      description: ['A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.', 'An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the location comes to rest in this alley flowery and colorful.'],
    },
    reviews: [reviews[1], reviews[3]],
  },
];

export const locationOffers: LocationOffers = [
  {
    location: LocationItem.Paris,
    offers,
  },
  {
    location: LocationItem.Amsterdam,
    offers,
  },
];
