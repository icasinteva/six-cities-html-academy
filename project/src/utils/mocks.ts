import { address, commerce, datatype, date, image, internet, lorem, name, random } from 'faker';
import { CITIES, MAP_ZOOM, OfferType } from '../const';
import { Offer, Review } from '../types/offer';
import { ReviewData } from '../types/review-data';
import { User } from '../types/user';

export const makeFakeUser = (): User => ({
  avatarUrl: internet.avatar(),
  name: name.firstName(),
  email: internet.email(),
  isPro: datatype.boolean(),
  token: random.alphaNumeric(32),
  id: datatype.number(),
} as User);

export const makeFakeOffer = (options?: { isFavorite?: boolean, isPremium?: boolean }): Offer => ({
  bedrooms: datatype.number({ max: 5 }),
  city: CITIES[datatype.number({ max: 6})],
  description: lorem.sentences(),
  goods: [...new Set(Array.from({ length: 8 }).map(() => commerce.product()))],
  host:  {
    avatarUrl: internet.avatar(),
    id: datatype.number(),
    isPro: datatype.boolean(),
    name: name.firstName(),
  },
  id: datatype.number(),
  images: Array.from({ length: 10 }).map(() => image.imageUrl()),
  isFavorite: options?.isFavorite ?? datatype.boolean(),
  isPremium: options?.isPremium ?? datatype.boolean(),
  location: {
    latitude: +address.latitude(),
    longitude: +address.longitude(),
    zoom: MAP_ZOOM,
  },
  maxAdults: datatype.number({ min: 1, max: 10}),
  previewImage: image.imageUrl(),
  price: +commerce.price(),
  rating: datatype.number({ min: 1, max: 5, precision: 0.1 }),
  title: lorem.sentence(),
  type: random.arrayElement([OfferType.Apartment, OfferType.Hotel, OfferType.House, OfferType.Room]),
});

export const makeFakeReviewData = (): ReviewData => ({
  comment: lorem.sentences(3),
  rating: `${datatype.number({ min: 1, max: 5, precision: 0.1 })}`,
});

export const makeFakeReviews = (quanatity: number): Review[] => Array.from({ length: quanatity }).map((_) => ({
  comment: lorem.sentences(),
  id: datatype.number(),
  rating: datatype.number({ min: 1, max: 5, precision: 0.1 }),
  user: makeFakeUser(),
  date: `${date.recent()}`,
}
));

export const makeFakeOffers = (quanatity: number): Offer[] => Array.from({ length: quanatity }).map((_) => makeFakeOffer());

export const makeFakeFavorites = (quanatity: number): Offer[] => Array.from({ length: quanatity }).map((_) => makeFakeOffer({ isFavorite: true }));
