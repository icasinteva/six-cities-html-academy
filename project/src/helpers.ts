import { LocationItem } from './const';
import { LocationOffers } from './types/offer';
import { locationOffers } from './mocks/offers';

export const getOffersByLocation = (offers: LocationOffers, activeLocation: LocationItem) => offers.find((offer) => offer.location === activeLocation)?.offers ?? [];

export const getFavorites = () => locationOffers.map(({ location, offers }) => ({ location, offers: offers.filter(({ favorite }) => favorite) })).filter(({offers}) => offers.length);

