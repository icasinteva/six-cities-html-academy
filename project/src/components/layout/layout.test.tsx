import { configureMockStore } from '@jedmao/redux-mock-store';
import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import HistoryRouter from '../history-route';
import Layout from './layout';
import { makeFakeFavorites, makeFakeOffers } from '../../utils/mocks';
import { getFavoritesByCity } from '../../services/helpers';

const mockStore = configureMockStore();
const fakeFavorites = makeFakeFavorites(10);

const store = mockStore({
  USER: {},
  OFFER: {},
  OFFERS: {
    offers: makeFakeOffers(10),
  },
  FAVORITES: {
    favorites: getFavoritesByCity(fakeFavorites),
  },
});

const history = createMemoryHistory();

describe('Component: Layout', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <HistoryRouter history={history}>
          <Layout />
        </HistoryRouter>
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });
});
