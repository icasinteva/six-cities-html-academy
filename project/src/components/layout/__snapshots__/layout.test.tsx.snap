// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Layout should render correctly 1`] = `
<div>
  <div
    class="page page--index page--gray"
  >
    <header
      class="header"
    >
      <div
        class="container"
      >
        <div
          class="header__wrapper"
        >
          <div
            class="header__left"
          >
            <a
              class="header__logo"
              data-testid="header__logo"
              href="/"
            >
              <img
                alt="6 cities logo"
                class="header__logo-link"
                height="41"
                src="img/logo.svg"
                width="81"
              />
            </a>
          </div>
          <nav
            class="header__nav"
            data-testid="header__nav"
          >
            <ul
              class="header__nav-list"
            >
              <li
                class="header__nav-item user"
              >
                <a
                  class="header__nav-link header__nav-link--profile"
                  data-testid="header__nav-link--profile"
                  href="/login"
                >
                  <div
                    class="header__avatar-wrapper user__avatar-wrapper"
                  />
                  <span
                    class="header__login"
                    data-testid="header__login"
                  >
                    Sign in
                  </span>
                </a>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </header>
    <main
      class="page-main page__main--index"
    />
  </div>
</div>
`;
